Design doc on reusing typechecking results (#17368)
dotnet/fsharp: docs/reusing-typecheck-results.md

|------------------------------------|--------|--------|-------|-------|-------|-------|-------|-------|
// main1
|Import mscorlib+FSharp.Core         |  0,2720|  0,2604|     96|      0|      0|      0|    365|     30|     <-- long
|Parse inputs                        |  0,3378|  0,0560|    108|      0|      0|      0|    372|     30|
|Import non-system references        |  0,3932|  0,0476|    127|      0|      0|      0|    372|     30|
|Typecheck                           |  0,9020|  0,5025|    164|      3|      2|      1|    456|     46|     <-- longest
|Typechecked                         |  0,9087|  0,0002|    164|      0|      0|      0|    456|     46|
// main2
|Write Interface File                |  0,9191|  0,0000|    164|      0|      0|      0|    456|     46|
|Write XML doc signatures            |  0,9280|  0,0000|    164|      0|      0|      0|    456|     46|
|Write XML docs                      |  0,9342|  0,0002|    164|      0|      0|      0|    456|     46|
// main3
|Encode Interface Data               |  0,9778|  0,0345|    164|      0|      0|      0|    456|     46|

> |Optimizations                       |  1,2312|  0,2463|    178|      2|      2|      1|    456|     46|     <-- long

Reviewer: ```suggestion
|Optimizations                       |  1,2312|  0,2463|    178|      2|      2|      1|    456|     46|     <-- This is long, but not relevant in the dev loop with run/debug/test cycles
```

|Ending Optimizations                |  1,2386|  0,0000|    178|      0|      0|      0|    456|     46|
|Encoding OptData                    |  1,2549|  0,0093|    178|      0|      0|      0|    456|     46|
|TailCall Checks                     |  1,2687|  0,0054|    178|      0|      0|      0|    456|     46|
// main4, main5
|TAST -> IL                          |  1,3657|  0,0883|    180|      0|      0|      0|    456|     46|

> |>Write Started                      |  1,3817|  0,0016|    180|      0|      0|      0|    455|     46|
> |>Module Generation Preparation      |  1,3909|  0,0002|    180|      0|      0|      0|    455|     46|
> |>Module Generation Pass 1           |  1,3987|  0,0015|    181|      0|      0|      0|    455|     46|
> |>Module Generation Pass 2           |  1,4266|  0,0216|    182|      0|      0|      0|    455|     46|
> |>Module Generation Pass 3           |  1,4353|  0,0024|    182|      0|      0|      0|    455|     46|
> |>Module Generation Pass 4           |  1,4423|  0,0004|    182|      0|      0|      0|    455|     46|
> |>Finalize Module Generation Results |  1,4515|  0,0003|    182|      0|      0|      0|    455|     46|
> |>Generated Tables and Code          |  1,4607|  0,0004|    182|      0|      0|      0|    455|     46|
> |>Layout Header of Tables            |  1,4663|  0,0001|    182|      0|      0|      0|    455|     46|
> |>Build String/Blob Address Tables   |  1,4750|  0,0013|    182|      0|      0|      0|    455|     46|
> |>Sort Tables                        |  1,4838|  0,0000|    182|      0|      0|      0|    455|     46|
> |>Write Header of tablebuf           |  1,4931|  0,0015|    182|      0|      0|      0|    455|     46|
> |>Write Tables to tablebuf           |  1,4999|  0,0000|    182|      0|      0|      0|    455|     46|
> |>Layout Metadata                    |  1,5070|  0,0000|    182|      0|      0|      0|    455|     46|
> |>Write Metadata Header              |  1,5149|  0,0000|    182|      0|      0|      0|    455|     46|
> |>Write Metadata Tables              |  1,5218|  0,0001|    182|      0|      0|      0|    455|     46|
> |>Write Metadata Strings             |  1,5297|  0,0000|    182|      0|      0|      0|    455|     46|
> |>Write Metadata User Strings        |  1,5369|  0,0002|    182|      0|      0|      0|    455|     46|
> |>Write Blob Stream                  |  1,5433|  0,0002|    182|      0|      0|      0|    455|     46|
> |>Fixup Metadata                     |  1,5516|  0,0003|    182|      0|      0|      0|    455|     46|
> |>Generated IL and metadata          |  1,5606|  0,0015|    182|      0|      0|      0|    455|     46|
> |>Layout image                       |  1,5713|  0,0018|    183|      0|      0|      0|    455|     46|
> |>Writing Image                      |  1,5799|  0,0009|    183|      0|      0|      0|    454|     46|
> |>Signing Image                      |  1,5871|  0,0000|    183|      0|      0|      0|    454|     46|

Reviewer: ```suggestion

```

|Typechecked                         |  0,6522|  0,0004|    144|      0|      0|      0|    378|     30|
// main2
|Write Interface File                |  0,6597|  0,0000|    144|      0|      0|      0|    378|     30|
|Write XML doc signatures            |  0,6661|  0,0000|    144|      0|      0|      0|    378|     30|
|Write XML docs                      |  0,6710|  0,0002|    144|      0|      0|      0|    378|     30|
// main3
|Encode Interface Data               |  0,7273|  0,0503|    154|      0|      0|      0|    378|     30|
|Optimizations                       |  0,8757|  0,1425|    172|      1|      1|      0|    378|     30|     <-- long
|Ending Optimizations                |  0,8815|  0,0000|    172|      0|      0|      0|    378|     30|
|Encoding OptData                    |  0,8899|  0,0024|    173|      0|      0|      0|    378|     30|
|TailCall Checks                     |  0,8990|  0,0025|    173|      0|      0|      0|    378|     30|
// main4, main5
|TAST -> IL                          |  0,9487|  0,0447|    176|      0|      0|      0|    378|     30|

> |>Write Started                      |  0,9664|  0,0023|    177|      0|      0|      0|    377|     30|
> |>Module Generation Preparation      |  0,9728|  0,0001|    177|      0|      0|      0|    377|     30|
> |>Module Generation Pass 1           |  0,9821|  0,0022|    177|      0|      0|      0|    377|     30|
> |>Module Generation Pass 2           |  1,0012|  0,0138|    178|      0|      0|      0|    377|     30|
> |>Module Generation Pass 3           |  1,0098|  0,0012|    179|      0|      0|      0|    377|     30|
> |>Module Generation Pass 4           |  1,0180|  0,0002|    179|      0|      0|      0|    377|     30|
> |>Finalize Module Generation Results |  1,0264|  0,0007|    179|      0|      0|      0|    377|     30|
> |>Generated Tables and Code          |  1,0344|  0,0003|    179|      0|      0|      0|    377|     30|
> |>Layout Header of Tables            |  1,0403|  0,0000|    179|      0|      0|      0|    377|     30|
> |>Build String/Blob Address Tables   |  1,0481|  0,0010|    179|      0|      0|      0|    377|     30|
> |>Sort Tables                        |  1,0543|  0,0000|    179|      0|      0|      0|    377|     30|
> |>Write Header of tablebuf           |  1,0608|  0,0003|    179|      0|      0|      0|    377|     30|
> |>Write Tables to tablebuf           |  1,0681|  0,0000|    179|      0|      0|      0|    377|     30|
> |>Layout Metadata                    |  1,0738|  0,0000|    179|      0|      0|      0|    377|     30|
> |>Write Metadata Header              |  1,0796|  0,0000|    179|      0|      0|      0|    377|     30|
> |>Write Metadata Tables              |  1,0875|  0,0000|    179|      0|      0|      0|    377|     30|
> |>Write Metadata Strings             |  1,0933|  0,0000|    179|      0|      0|      0|    377|     30|
> |>Write Metadata User Strings        |  1,1016|  0,0001|    179|      0|      0|      0|    377|     30|
> |>Write Blob Stream                  |  1,1076|  0,0001|    179|      0|      0|      0|    377|     30|
> |>Fixup Metadata                     |  1,1135|  0,0004|    179|      0|      0|      0|    377|     30|
> |>Generated IL and metadata          |  1,1208|  0,0009|    179|      0|      0|      0|    377|     30|
> |>Layout image                       |  1,1269|  0,0010|    179|      0|      0|      0|    377|     30|
> |>Writing Image                      |  1,1347|  0,0003|    179|      0|      0|      0|    376|     30|
> |>Signing Image                      |  1,1402|  0,0000|    179|      0|      0|      0|    376|     30|

Reviewer: ```suggestion

```

|Typechecked                         |  3,1219|  0,0002|    198|      0|      0|      0|    712|     45|
// main2
|Write Interface File                |  3,1280|  0,0000|    198|      0|      0|      0|    712|     45|
|Write XML doc signatures            |  3,1363|  0,0000|    198|      0|      0|      0|    712|     45|
|Write XML docs                      |  3,1435|  0,0002|    198|      0|      0|      0|    712|     45|
// main3
|Encode Interface Data               |  3,1803|  0,0296|    198|      0|      0|      0|    712|     45|
|Optimizations                       |  8,9949|  5,8049|    216|     43|     42|      2|    457|     45|     <-- longest
|Ending Optimizations                |  9,0015|  0,0000|    216|      0|      0|      0|    457|     45|
|Encoding OptData                    |  9,0090|  0,0010|    216|      0|      0|      0|    457|     45|
|TailCall Checks                     |  9,0190|  0,0013|    216|      0|      0|      0|    457|     45|
// main4, main5
|TAST -> IL                          |  9,0463|  0,0210|    217|      0|      0|      0|    458|     46|

> |>Write Started                      |  9,0586|  0,0018|    217|      0|      0|      0|    457|     46|
> |>Module Generation Preparation      |  9,0655|  0,0001|    217|      0|      0|      0|    457|     46|
> |>Module Generation Pass 1           |  9,0719|  0,0005|    217|      0|      0|      0|    457|     46|
> |>Module Generation Pass 2           |  9,0820|  0,0031|    218|      0|      0|      0|    457|     46|
> |>Module Generation Pass 3           |  9,0894|  0,0005|    218|      0|      0|      0|    457|     46|
> |>Module Generation Pass 4           |  9,0968|  0,0002|    218|      0|      0|      0|    457|     46|
> |>Finalize Module Generation Results |  9,1032|  0,0003|    218|      0|      0|      0|    457|     46|
> |>Generated Tables and Code          |  9,1120|  0,0005|    218|      0|      0|      0|    457|     46|
> |>Layout Header of Tables            |  9,1189|  0,0000|    218|      0|      0|      0|    457|     46|
> |>Build String/Blob Address Tables   |  9,1276|  0,0006|    218|      0|      0|      0|    457|     46|
> |>Sort Tables                        |  9,1339|  0,0000|    218|      0|      0|      0|    457|     46|
> |>Write Header of tablebuf           |  9,1430|  0,0001|    218|      0|      0|      0|    457|     46|
> |>Write Tables to tablebuf           |  9,1503|  0,0000|    218|      0|      0|      0|    457|     46|
> |>Layout Metadata                    |  9,1588|  0,0000|    218|      0|      0|      0|    457|     46|
> |>Write Metadata Header              |  9,1656|  0,0000|    218|      0|      0|      0|    457|     46|
> |>Write Metadata Tables              |  9,1731|  0,0000|    218|      0|      0|      0|    457|     46|
> |>Write Metadata Strings             |  9,1819|  0,0000|    218|      0|      0|      0|    457|     46|
> |>Write Metadata User Strings        |  9,1897|  0,0000|    218|      0|      0|      0|    457|     46|
> |>Write Blob Stream                  |  9,1976|  0,0000|    218|      0|      0|      0|    457|     46|
> |>Fixup Metadata                     |  9,2033|  0,0002|    218|      0|      0|      0|    457|     46|
> |>Generated IL and metadata          |  9,2102|  0,0015|    218|      0|      0|      0|    457|     46|
> |>Layout image                       |  9,2195|  0,0025|    218|      0|      0|      0|    457|     46|
> |>Writing Image                      |  9,2272|  0,0006|    218|      0|      0|      0|    456|     46|
> |>Signing Image                      |  9,2338|  0,0000|    218|      0|      0|      0|    456|     46|

Reviewer: ```suggestion

```

All the steps starting with a pipe are just children of writing .NET Binary, they are not helpful in this doc as they stand.

Author: Fair, I just copypasted --times, this is a redundant part indeed.

--------------------------------------------------------------------------------------------------------
```

The phases taking the longest are marked in the right. Those are assembly import, type check, optimization and IL writing.

## Implementation plan

The conclusion from the above is that there is a lot of potential in caching - at the same time, implementing things to the full potential right away will require serious changes in the compiler code, which will be dangerous, hard to test and review. Therefore, I propose to implement the feature in stages where each stage brings gains in some scenarios.

**Stage 1 - force-gen and compare typechecking graphs**

We already create the TC graph in `main1` and we are able to dump it (in Mermaid) via the compiler flags.

So we can do the following:
- force-gen and save the graph
- append all the extra compilation info (basically, the compilation arguments string)
- before doing the retypecheck, detect if there is any change to above:
  - if so, skip the retypecheck
  - otherwise, apply some algorithm to get the graph diff. The result should be the list of files for retypecheck (just ignore it for now)

> This will allow to skip a huge part of the compilation in the scenarios like reopening Visual Studio for a project when nothing has changed.

Reviewer: ```suggestion
This will allow to skip a huge part of the compilation in the scenarios like reopening Visual Studio for a project when nothing has changed.

This step is largely technical without big observable benefits, but will require the necessary msbuild hooks to communicate the intermmediatefiles folder, add time-based and hash-based cache invalidation logic, and involve necessary tests which will include the `clean` and `rebuild` tests to make sure the cache be easily invalidated on demand.
```

**Stage 2 - skip retypecheck for some files**

> In `main1`, we do unpickling (= deserializing) and pickling (= serializing) of the code. This currently happens on the module/namespace basis. 

Reviewer: ```suggestion
We will fully implement pickling and unpickling of all typechecked files (`CheckedImplFile` per file as well as all other outputs of `main1` which cannot be cheaply re-created like topAttrs and CcuThunk). 

Parts of the pickling/unpickling can reuse the primitives already existing in TypedTreePickle and built upon them.
This is likely the biggest amount of work expected.
```

So here we can do the following:

> - extend the pickler logic so that it can work on the file basis

Reviewer: ```suggestion

```

That is not the problem at all. This is nature of the work to be done, to pickle really everything which is needed to avoid the full `Typecheck` step and replace it with a step that does not need to `read >> lex >> parse >> typecheck` the input file at all.

Author: Well that wasn't meant to be a problem, but a task :)

- save acquired typecheck information to the intermediate folder after the typecheck
- for the files not needing retypecheck as per graph diff, skip the phase, instead restore the typecheck info

This will hugely benefit the scenarios when only the edge of the graph is affected (think one test in a test project).

> **Stage 3 - reduce signature data generation**
> 
> In `main3`, we encode signature data in the assemblies.
> 
> Since we'll have typecheck graph at this point, we can detect its "edges" - files which nothing depends on. Those don't need to include signature data, hence we can skip the encoding step for them.
> 
> Here, the gain largely depends on the type of the project, but this would be anyway a good move in the direction of the compiler less relying on signatures for its operations.
> 

Reviewer: ```suggestion
**Stage 3 - reduce signature data generation**

In `main3`, we encode signature data in the assemblies.

Since we are addressing run/debug/test scenarios, a lot of projects can avoid signature data at all. Reason is, signature data is only needed for cross-project compilation. This could be detected automatically for certain types of projects.

For libraries, the resource blob (a byte array) would have to be split into per-file data, and logic added to recombine the splits with freshly typechecked information into a single binary resource back.

```

This does not make much sense.
In order of importance:
- For edge projects (webapp, exe file, test project), signature files can be disable alltogether to save time. This should be the first focus point, since it saves the most and does not need extensive logical change.
- 
- For non-edge project (=libraries), the signature data can be persisted and rehydrated, at the cost of additional logic to make it fit after incremental changes. (i.e. assembling 99/100 pieces if 1 file changed, but then putting it together with 1 re-typechecked file consistently in a single resource block which is still valid and legit)
- The last sentence does not make much sense in the context of this design doc

Author: Yeah, I've probably misunderstood this piece during our discussion, addressed now.

**Stage 4 - reuse import data**

In `main1`, we restore imported assemblies to get their IL (e.g. system assemblies).

So here we can:

> - serialize and cache imported IL on the per-assembly basis

Reviewer: ```suggestion
- serialize and cache imported IL (to be evaluated and benchmarked if per-assembly or per assembly-block, such as all System.*.dll into a single persisted file)
```

Author: Good point, worth checking out.

- apply this for the cross-project case, to not reimport IL for repeating assemblies but instead restore it from the cache

> This will be a smaller gain for any particular project but a big accummulated one for large multi-project solutions.

Reviewer: ```suggestion
This will be a smaller gain for any particular project but a big accummulated one for large multi-project solutions.

This step includes adding a cross-project intermmediate file location to be coordinated with msbuild properties.
```

---

> title: Reusing compilation results

Reviewer 2: No, it's still typecheck results, but in compiling scenarios, not editor.

category: Compiler Internals
categoryindex: 200
index: 42
---

# Reusing compilation results between compiler and tooling runs

Caching and reusing compilation results between compiler and tooling runs will be an optimization technique aimed at improving the performance and efficiency of the F# development experience.

## Motivation

- **Performance**. Things like recomputing type information for large projects are time-consuming. By caching the results, subsequent compilation can skip some steps, leading to faster builds.

- **Better IDE cold start**. IDEs can provide faster IntelliSense, code navigation, and other features if they can quickly access cached compilation information.

> - **Error reduction**. Consistent compilation results help avoid scenarios where slight differences between runs could cause different errors or warnings to be produced.

Reviewer: What are the specific examples?
How would the feature help at them?

Without real examples, it would be better to have this bullet point excluded to avoid confusion of what this can really bring.

Author: Removed also

### Example real-world scenarios

The optimization will benefit the most in large and not very interdependent projects. For instance, a 100-files test project where a single test is changed. MSBuild will mark the project for recompilation but thanks to the improvement only one file would be recompiled in the full-blown manner.

> The technique should also help when MSBuild generally incorrectly marks projects for rebuild. The probably doesn't happen often though.

Reviewer: It's done on timestamps, which this feature will likely be doing the same.
=> better to delete this

### What should be done

**Stage 1 - omit retypecheck for plain rebuild**

That is, after running "fsc test.fs", rerunning `fsc test.fs` will omit type-checking step. This should work for multiple files (project) as well. Everything will be happening in `main1` - after parsing, we check the cache and deserialize the typecheck info if available, passing it to `main2` right away. If not, we do the normal type check, caching the results at the end of it, prior to passing them to the next stage. This will speed up scenarios like tinkering with plain "fsc.exe" or project rebuild when the results are incorrectly invalidated.

**Stage 2 - minimize retypecheck in slightly modified projects**

That is, when rebuilding the project after modifying an independent file, only the modified file will be retypechecked, for others the check will be omitted. This will handle many target scenarios for the feature, like modifying one test in a test project.

**Stage 3 - omit reresolving assemblies**

> That is, for either of the scenarios above, skip the `Import mscorlib+FSharp.Core` step. This logic is also done in `main1`, prior to parsing. Same ideas apply - if cached, deserialize the information - otherwise, follow the normal process and then cache the results. The data should be stored separately from the typecheck info, so that the information be reused partially.

Reviewer: Consider a cross-project storage of this step, it is highly repeated (e.g. for all System.* libs).
Beyond mscorlib+FSharp.Core, there are also all the other dependencies coming from package and project references.

Author: Yep, good point.

**Stage 4 - omit reoptimization**

> That is, for everything that doesn't need to be retypechecked, don't do reoptimization either. This happens in the `main3` phase of compilation. 

Reviewer 2: This makes little sense. With small changes we will need to run optimisation either way. With no changes, nothing will be rebuilt. Already isn't, there are up to date checks in place already.

Reviewer: I would go even further - do not bother with opt phase at all.
This is about improving inner dev loop by making it fast - turning off optimizations (e.g. by using DEBUG) is something users should be already doing.

I think the rare occasions of debugging an optimization-related issue will be fine with running the full optimization phase.

Reviewer 2: Yeah, that makes even more sense. I think we should focus on debug config scenarios, as it's the most common one for the dev loop. And in debug we don't optimize.

Author: Alright, removed for brevity and clarity

### What else can be done

> **Omit rewriting .NET binaries**

Reviewer 2: It'll likely won't be done already.

> This means that if _for some reason_ the recompilation is triggered but there is nothing to retypecheck, reuse the binaries. This shouldn't be hard to do (just writing and reading existing files) - but it's unclear how often this happens in real life.

Reviewer: msbuild already does that if no input file changes.
For sake of simplicity and avoiding confusion, I think it would be better to delete this paragraph alltogether.

Author: Alright, removed for brevity and clarity.

**Omit all the other steps**

> If we have universal caching mechanism (see below), we can apply it to other steps (e.g. parsing). Those steps typically don't take much time each on their own, but can somewhat accummulate and maybe even peak in some edge cases. 

Reviewer: If an input file is modified, both typecheck data as well as parsing data has to be invalidated.
I do not really see a benefit there - rather cost of doing additional file writes, which will be invalidated by the same source of changes.

I would also recommend to just delete this paragraph here.

Author: Alright, removed for brevity and clarity.

**Integrate this into tooling**

> We can plug the feature into the tooling. This would lead to time savings in many more developer scenarios. However, this is not trivial and the tooling layer of the feature would need to do extra checks and invalidation calls, e.g. when the tooling and the compiler use different versions of the compiler service.

Reviewer: But language services for tooling already do have this, I think this paragraph is better to be deleted to avoid confusion.

Author: Alright, removed for brevity and clarity.

> **Omit recompilation for insignificant file changes**

Reviewer 2: In the current compiler design, I don't see this happening honestly, I would put it into separate group of "might be useful, but not currently realistic".

Author: Alright, removed for brevity and clarity.

That is about getting smart after getting first parts of the compilation results. If we know that the code is IL-same, we can skip everything and reuse the binaries. This would be useful in scenarios like adding comments or removing unused opens. However, this might present a **hard** (albeit interesting) challenge - things like figuring out AST diffs seem to be unexplored yet.

## Open questions

### How to cache data?

> We can store the data in the intermediate build directory (e.g. `obj` in case of `dotnet build`). The usual way for caching would be to serialize + deserialize information.

Reviewer: How?
What and how many files will be in the directory?

(compiler flags, dependency graph, data from system imports...)

Reviewer: E.g. a sample project:

A.fsi
A.fs
B.fs

However, serialization of the compiler data is non-trivial:

> - we don't have a **full** out-of-the-box solution in the repo (some parts are available, see below)

Reviewer 2: Built in pickler/unpickler should be suitable for majority of things. We already use it to encode signature data.

Author: Yes, looks like that, eventually I think there is no need to reinvent the wheel here.

- well-known serialization libraries (like `System.Text.Json`) won't work right away, because data is internal and libs can't handle some F# type easily (e.g. DUs)

> - we can't use F#-specific libraries (like [this](https://github.com/fsprojects/FSharp.Json) or [this](https://github.com/Tarmil/FSharp.SystemTextJson)) because they are not "official" enough

Reviewer 2: We can't use them because we don't take any dependencies, not because they're "not official".

Author: Didn't mean this in a "bad" way, reformulated.

It should be possible to hack things around, but the more compiler data we want to serialize, the harder it will be. Hence I see 2 ways to go:

**1) Start with leveraging some existing elements and then see.** 

> We have a mechanism for writing signature files. We could force generating signatures for fresh compilation, save them and then "feed" to the recompilation process. This will speed up typechecking, saving a lot of time on type inference. We also seem to have a mechanism for encoding optimization data. We can apply the same idea here - force saving it for the first run and then utilize for reruns. This won't help us with everything but **might be** good enough. Also, the things we'll be left with **might be** easier to hand-serialize.

Reviewer: This paragraph is too vague, it would be a lot clearer if it used actual terms from the compiler codebase.
- Which types would be missing considering the combined scope of signature+optdata pickler?

Author: Remove for now, pickling + unpickling might be enough.

**2) Create the universal serialization + deserialization mechanism right away and use it for everything**

> This can take inspiration from existing mechanisms and random libraries on the internet. Here we'll be able to unlock the full potential of data caching, yet it will also mean a bigger investment in the beginning. Also, the caching procedure would be more uniform here - we can probably store whatever we can per each file, in the same format. 

Reviewer: I think its better for everyone if this paragraph isn't even there.
Universal binary serialization of arbitrary types and object graphs composed of them - people tried before.

Author: Removed

### Cache invalidation

It's clear that we should be as defensive as possible here, cancelling further information reuse when suspecting that it might not match the fresh recompilation info.

**Checking source files**

To understand the source file is the same we can:
1) Check the timestamp
2) Hash the content and compare the hashes

The former might lead to problems with things like time zone changes or manual file metadata modification. The latter might lead to problems in case of hash collisions. All of those scenarios are probably very unlikely, we might just look at what MSBuild does.

**Checking auxiliary information**

Besides tracking the source files, we should track:
- compiler version
- compiler flags 
- ?

**Checking dependencies**

> This is related to the minimization of retypecheck in slightly modified projects. If only one file is changed, how do we know that nothing depends on that?

Reviewer 2: We don't, that's why we also should save graph TC information. But this might also be not enough, since graph can change even if we save one file. That's why we should for now assume that all files below changed.

Reviewer: With graph from last run + graph from new run, you can determine the potential scope of invalidated files (a union of the edges).

Not doing this will significantly decrease usefulness of this features for projects like tests (where the graph has practically no edges).

Reviewer 2: Additionally we should also see how different graphs for projects with signature data and data without it, and see what's the difference will look like for both.

Author: Good point about actually caching the TC graph, added.

Graph-based typechecking mechanism should help here, however this also means we might need to store some additional data to keep track of things.

### Data propagation

> Since the compilation steps we want to omit happen in different phases of compilation, I see 2 approaches to data propagation:

Reviewer: I would recommend to fully ignore optimization data in this improvement for inner dev loop, it's not a very relevant scenarios (yes, I know there are small exceptions).

What I am not seeing here:

You will need both the old graph + new graph to determine which files are invalidating each other. How this affect phasing? Will source files be always read, if not, how will the compiler work without having source file available?

How does information propagate between .fs and .fsi files, how will their consistency work if one is cached and the other not?

Author: Why wouldn't we read the source files? As in, why would we want to change this for this feature?

**1) Cache and restore data in each phase.**

For example, independently restore typecheck data and optimization data just before doing the respective jobs. This would lead to redundant checks (we don't need to reoptimize if we don't retypecheck), hence some code clutter, but would keep the general code structure same. We'll also have to leak the data a bit (invalidate optimization cache if we invalidate typecheck cache).

**2) Restore at the beginning, cache at the end.**

This would mean doing all the data restoring either in `main1` or even in a new phase dedicated to just that. Then the information would be gradually collected and propagated to the `main6`. This will avoid unnecessary checks but will also mean modifying the `Args` data and hence significantly intervening into the current compiler logic. 

## Testing and benchmarking

> We should have a feature flag for this: applying it to current typechecking tests shouldn't make any difference in results. Unit tests should test that restored cached results + reusing them should be equivalent to fresh typecheck results.

Reviewer: Dedicated compiler switch, not a feature flag.

Author: Yep, fixed.

---
title: Reusing typecheck results
category: Compiler Internals
categoryindex: 200
index: 42
---

# Reusing typecheck results between compiler and tooling runs

Caching and reusing typecheck results between compiler and tooling runs will be an optimization technique aimed at improving the performance and efficiency of the F# development experience.

### Why is it important

- **Performance**. Recomputing type information for large projects is time-consuming. By caching the results, subsequent runs can skip this step, leading to faster builds.

> - **IDE responsiveness**. IDEs can provide faster IntelliSense, code navigation, and other features if they can quickly access cached type information.

Reviewer: I think this motivational area gets the important aspects wrong.
IDE already does keep and reuse both parse results as well as typecheck results, and TransparentCompiler made a step in making the invalidation even less aggressive. IMO this design doc should do not attempts at tooling reuse, because it would just steal focus and repeat information which has been already done. What is a big deal and should be considered is using tooling cache's as a source of information - but that would need to be guided by strict checks for exact versions and project settings, which is unlikely to be met in current state of the F# extension.

The scenarios expected to get **massive** benefits are not covered here, but yet should be the scenarios around which the feature ergonomics will be designed around:

1. Have a large test project with deterministically independent fiels (e.g. verified independence via graph checking's algorithm), run tests. Adjust a single line of code in a single test and rerun that test again.
2. Have a web app, run it from VS. Add a parameter to a single httphandler at the bottom of the project and rerun the same webapp.
3. Have a large console app under debugger. Modify a private function in 1 implementation file in the middle of the project, without affecting intra-file and intra-project signatures, and repeat debugging for that console app.

Author: Good point also. I reformulated the motivation and added some real world scenarios, thanks for the inputs.

- **Error reduction**. Consistent typechecking results help avoid scenarios where slight differences between runs could cause different errors or warnings to be produced.

### General considerations

> - The technique will be implemented separately for the compiler and tooling, for better granularity and trackability.

Reviewer 2: In addition to that, tooling and compiler can be using different version of compiler service, and technically results might be incompatible

Reviewer 3: Yeah this will be a big hurdle (repeat analyzers discussions)

Author: Yes, we need to keep the track of the compiler version, noted.

> - The technique will be implemented separately for the compiler and tooling, for better granularity and trackability.

Reviewer: I would rather twist the argumentation around.
"The feature will get the best benefits if data is reused between tooling and compiler runs. However, for practical reasons (versioning, project settings, runtime difference) it will not be implemented from the start".

i.e. the separation is rather a technical limitation at the moment, but should not become part of the axioms that affect the overall design.

If those limits are lifted (e.g. tooling using the same F.C.S. as building), the potential is even bigger - running tests or starting a debugging session could be utilizing a fully cached typecheck result (and consequently imports, parsing,..)

Author: Agree. I decided to not focus on tooling for now since there is a lot to decide on even in the compiler. I put the tooling support to the followups.

Reviewer: I did not mean the area of biggest potential as supporting tooling at all, it was meant as supporting compilation (build, test, run, debug,..) with information collected **from** tooling.

That way, starting tests could feel almost instantaneous, because the type checking would be already happening while user is making changes.

> - Caching data will likely be unsophisticated, the harder part will be to understand what data can be reused and how.

Reviewer 2: I would say choosing format for serialisation might be challenging, since we need it to be iterative

## Detailed design: compiler layer

### Caching

> - Cache data to the "obj" folder, next to other intermediary compilation information.

Reviewer 2: I would rename it to intermediate output folder (which will be coming from msbuild or whoever is invoking compiler).

Author: Fair enough, renamed.

- Do this during Phase 6 of the compilation, next to writing binaries (`main6`).
- Store the data either in binary or in text, as with other things in "obj". The decision can maybe be made based on the speed of serialization/deserialization of different formats.

### Reusing

- Start reading data asynchronously in the beginning of compilation, around importing assemblies (beginning of `main1`).
- Await the data just before the type check (middle of `main1`).
- If there is no data, do type check as usual.

> - If there is some data, see what can be reused.

Reviewer 2: This will need to be expanded.

> ### What data to store?

Reviewer: Missing aspects:

- Storage granularity and design discussion about it's tradeoffs:
    - File per source file: Makes timestamp comparison fast and straightforward. Makes reading slower if all/most files are eventually read.
    - File per project: Must eagerly load even when cache is not used
    - File per project with custom binary header and lazy reading: Added complexity of custom binary format and seeking within a single file
    - Utilizing an existing format like .zip, .tar.gz or possibly even.sqlite: Depends on data. .sqlite might have its place for structured relational data like symbol usages or semantic tokenization (i.e. the "IDE cold start" situation), but not so much added benefit for binary blobs in regular builds.

- Start with storing full typecheck data.

> - Later parsing data can be added if we find a good scenario for having just that. 

Reviewer: Lexer and parser are really fast, the ROI here would be rather at the end of the list.
Unlike imports (e.g. single blob as opposed to reading dozens of files) or dependency-graph (will decrease cascading effect of invalidations).

Author: Fair enough. I think we might still want to do it eventually, if we have a general caching mechanism. But it's true that the returns diminish here - put it to the list of possible followups.

Author: Removed for now for brevity here.

### How to reuse data?

- Be defensive - if there's not 100% guarantee something can be reused, better just invalidate it.

> - We need to understand how to reuse data **partially**. Probably can be deduced based on AST diff in parsing?

Reviewer 3: Yeah this will be tricky. There's gotta be a way to handle simple things like `Adding a // comment` that doesn't force a re-check of a solution.

Reviewer 2: That'll be an interesting challenge, especially when caching source/ast and then adjusting ranges if needed, etc.

Reviewer: I don't think even mentioning "partial" and "AST diff" fits here.

It's easily the most difficult bit of this document, yet the feature would be extremely useful without it.
If tooling (i.e. something done as you type each individual letter) did not have this for 15 years, why complicating this feature with it?
I would suggest for the doc should be realistic in what is achievable in a single feature timeframe.

Author: Fair enough. I didn't remove this thought entirely, but put it to the end of the list of possible followups.

## Implementation plan

The conclusion from the above is that there is a lot of potential in caching - at the same time, implementing things to the full potential right away will require serious changes in the compiler code, which will be dangerous, hard to test and review. Therefore, I propose to implement the feature in stages.

**Stage 1 - force-gen and use signatures**

> We can start with always generating signature files and using them for typecheck in cases when matching implementation files don't change. The affected phase will be `main2`.

Reviewer: I don't think this would serve well as a "phase" in the overall goal meeting the motivation.
For me, the motivation always was "For a minimalistic change within 1 function, in a single file, in a huge project without many intra-dependencies, we should get **significantly** faster".

I don't think the proposal gets there - in the end it will be equivalent to having a project with signature files, which e.g. this repository already has and therefore this repo would not see any benefit in case of changing implementation details (i.e. a change which does not need changing a .fsi file).

Considering the amount of real-world broad testing of the signature text-generation, introducing this as a background optimization is too risky. Very likely, errors coming from mismatch between generated file and the underlying source will be fixable only be people who understand .fsi files and their different syntax in the first place, making it extremely expert-oriented.

This together with foreseen potential leads me to believe that this is not a stage (i.e. not something that can act as useful evolution step), but rather a distraction.

Any strong solution must aspire to eliminate most of the parsing and type checking time, not replace it with something roughly equivalent (duration-wise).
If I could propose, I would suggest to make the "stages" based on robustness against various kinds of changes within a project (e.g. stages of utilizing intra-file dependencies via: linear ordering `>` parsed graph `>` typechecked graph > detecting impl-only changes not affecting other files)

Author: It's fair that FCS itself will not benefit too much from this particular change, since it's already full of signatures.

That said:
1) Even "this repo" contains a lot of code not backed with signature files
2) We don't promote sigfiles, so there might be other big repos without sig files at all to heavily benefit from this
3) If this gets properly tested, eventually it can be a step towards getting rid of sigfiles, which will be a good thing

Help me understand your proposal here. Do you mean that force generating signatures should be:
- a separate task altogether?
- later in this impl plan (e.g. after caching opt data)?
- together with later steps right away? (e.g. with caching opt data)?

Reviewer: And yes, if persistence of type-checked data works correctly and cache invalidation will be eventually taught to navigate parsed dependency graphs, I do believe this might be a good motivation to stop using .fsi files just for the sake of compilation speed benefits (as it is now).

It still might have its place for a handful of selected files as API guard, but that is not something to be kept (nor generated) for the entire project.

Author: > None of it, I do not think generating text-based signatures on the background without user's intention is good at all.

Well that's what I meant by "separate task" then. No okay I am not dying to dive into elaborating signature generation logic in the compiler, I was just under the impression this was something we agreed on. It's fair that if we achieve significant perf improvements with other means, this will also minimize the need of explicit signature files altogether.

Would you then start with caching sigData + optData right away, or do you think comparing typechecked graphs and acting on that is a better first step, taking into account both the potential gains and implementation complexity?

And @vzarytovskii what do you think?

Reviewer: Comparing typechecked graphs is for cache invalidation, in order to invalidate smallest set of files possible.

Persisting and rehydrating typecheck results (which is more than just sigData + optData in their today's forms, but enlarging the set of values to be persisted is the real work to be done by such a feature) is IMO the first step to make anything work. Even when invalidation would work by mere linear ordering initially, that would still be a good starting point.

Author: Addressed in the new plan.
